src/spend/secret.rs

```diff
@@ -0,0 +1,143 @@
+use super::public::PublicKey;
+use super::stealth::StealthAddress;
+
+use crate::sponge;
+use crate::{JubJubScalar, ViewKey};
+
+use dusk_jubjub::GENERATOR_EXTENDED;
+
+use std::fmt;
+
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+use rand::{CryptoRng, RngCore};
+use sha2::{Digest, Sha256};
+
+/// Secret pair of `a` and `b`
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SecretKey {
+    a: JubJubScalar,
+    b: JubJubScalar,
+}
+
+impl Default for SecretKey {
+    fn default() -> Self {
+        SecretKey {
+            a: JubJubScalar::random(&mut rand::thread_rng()),
+            b: JubJubScalar::random(&mut rand::thread_rng()),
+        }
+    }
+}
+
+impl SecretKey {
+    /// This method is used to construct a new `SecretKey` from the given secret
+    /// pair of `a` and `b`.
+    pub fn new(a: JubJubScalar, b: JubJubScalar) -> Self {
+        Self { a, b }
+    }
+
+    /// Gets `a`
+    pub fn a(&self) -> &JubJubScalar {
+        &self.a
+    }
+
+    /// Gets `b`
+    pub fn b(&self) -> &JubJubScalar {
+        &self.b
+    }
+
+    /// Generate a `sk_r = H(a · R) + b`
+    pub fn sk_r(&self, sa: &StealthAddress) -> JubJubScalar {
+        let aR = sa.R() * self.a;
+        let aR = sponge::hash(&aR);
+        aR + self.b
+    }
+
+    /// Deterministically create a new [`SecretKey`] from a random number
+    /// generator
+    pub fn from_rng<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let a = JubJubScalar::random(rng);
+        let b = JubJubScalar::random(rng);
+
+        SecretKey::new(a, b)
+    }
+
+    /// Derive the secret to deterministically construct a [`PublicKey`]
+    pub fn public_key(&self) -> PublicKey {
+        let A = GENERATOR_EXTENDED * &self.a;
+        let B = GENERATOR_EXTENDED * &self.b;
+
+        PublicKey::new(A, B)
+    }
+
+    /// Derive the secret to deterministically construct a [`ViewKey`]
+    pub fn view_key(&self) -> ViewKey {
+        let B = GENERATOR_EXTENDED * &self.b;
+
+        ViewKey::new(self.a, B)
+    }
+}
+
+impl From<&SecretKey> for [u8; 64] {
+    fn from(pk: &SecretKey) -> Self {
+        let mut bytes = [0u8; 64];
+        bytes[..32].copy_from_slice(&JubJubScalar::from(pk.a).to_bytes()[..]);
+        bytes[32..].copy_from_slice(&JubJubScalar::from(pk.b).to_bytes()[..]);
+        bytes
+    }
+}
+
+impl From<&[u8]> for SecretKey {
+    fn from(bytes: &[u8]) -> Self {
+        let mut hasher = Sha256::default();
+        hasher.input(bytes);
+        let bytes = hasher.result();
+
+        let mut seed = [0u8; 32];
+        seed.copy_from_slice(&bytes[..32]);
+
+        SecretKey::from_rng(&mut StdRng::from_seed(seed))
+    }
+}
```

Comment:
Is this necessary?

Reply:
Care to elaborate?

Reply:
It might be confusing since it can seem that it's intended to read bytes and from there get the key (as if it was a `Read` trait impl). If you want to get a random `SecretKey` just use the `from_rng` method.

Reply:
IMVHO it's not confusing because in order to read bytes to generate a `SecretKey` you don't do that from a slice of bytes. As we do for all our other types, you do it from a fixed size array.

In this case, it's not about having a random `SecretKey`, is about generating a `SecretKey` by arbitrary bytes. So, as you can see, the scope is different.

If your question is _"Do we have the need to generate a `SecretKey` from arbitrary bytes?"_ I will delegate the question to Togh. The original implementation has this feature, so I bring it with me in Dusk PKI.

Reply:
If the original impl has this feature I said nothing. Even if it looks or not confusing to me, the specs are the specs. So this LGTM therefore.